src/core/inductive.ml

```diff
@@ -0,0 +1,292 @@
+              in
+              if acc < curr_int then
+                aux q curr_int
+              else
+                aux q acc
+        in
+        let res = aux l (-1) in
+        e ^ (string_of_int (res+1))
+  in
+  let acc1, acc2 = aux_2acc (i::l) [] [] in
+  (get_prefix acc1 "x", get_prefix acc2 "p")
+
+(** [gen_ind_typ_codom pos ind_sym codom s] assumes that the type of [ind_sym]
+    is of the form [Π(i1:a1),...,Π(in:an), TYPE]. It then generates a [tbox]
+    similar to this type except that [TYPE] is replaced by
+    [codom [i1;...;in]]. The string [s] is the prefix of variables' name. *)
+let gen_ind_typ_codom : popt -> sym -> (tbox list -> tbox) -> string -> tbox =
+  fun pos ind_sym codom s ->
+  let rec aux : tvar list -> term -> tbox = fun xs a ->
+    match Basics.get_args a with
+    | (Type, _) -> codom (List.rev_map _Vari xs)
+    | (Prod(a,b), _) ->
+        let (x,b) = Basics.unbind_name b s in
+        _Prod (lift a) (Bindlib.bind_var x (aux (x::xs) b))
+    | _ -> fatal pos "The type of %a is not supported" pp_symbol ind_sym
+  in aux [] !(ind_sym.sym_type)
+
+(** [fold_cons_typ pos codom get_name build_rec_hyp domrec dom ind_sym
+    cons_sym f_rec f_not_rec acc s] generates some value iteratively by going
+    through the structure of [cons_sym.sym_type].
+    We introduce some notations:
+      - A product Π([x] : [a]), [b]
+      - [xs] stores each variable of the intial product during the computation
+      - [ts] is the arguments of the current symbol
+      - [rec_hyp] is the current recursive hypothesis
+      - [next] is the result of the recursive call
+      - [res] is a computation built incrementally
+    Now, we can describe each argument of the function [fold_cons_typ]:
+      - The argument [pos] is the position of the command inductive where the
+      constructors were defined.
+      - On the codomain, the function [codom ts xs res] is called on the
+      arguments to which [ind_sym] is applied [ts], the variables of the
+      products [xs] (in reverse order) and a computation built incrementally
+      [res].
+      - In case of a product, the function [get_name b i] splits the product
+      [b] and gives a name to the variable if it has none, thanks to the
+      number [i].
+      - In case of recursive occurrences, the function [build_rec_hyp ts x]
+      builds the recursive hypothesis associated, and then the function
+      [domrec a x rec_hyp next] is applied to conclude the building.
+      - Otherwise, the function [dom a x next] is called.
+      - If you would like to store a temporary result, the initial value is
+      [acc], and you can change this value in the recursive case with the
+      function [f_rec res x rec_hyp], and on the other case with the function
+      [f_not_rec res x].
+      - The string [s] is the prefix of variables' name. *)
+let fold_cons_typ (pos : popt) (codom : term list -> 'b list -> 'a -> 'c)
+      (inj_var : 'b list -> tvar -> 'b)
```

Comment:
Isn't it possible to replace in the type of inj_var `'b list` by `tvar list`?

Reply:
Idem in the type of aux.

Reply:
Question: can't we replace `'b` by `tvar` everywhere?

Reply:
Not really because of the instance of fold to create rewriting rules. In this case, `'b` is instancied by `p_term`.

Reply:
Currently yes but perhaps you can change the functions a little bit...

Reply:
I tried to do this but I think it's not possible: How create a tvar that can be transformed by `P_wild`? See `appl_wild`.

Reply:
Maybe it's possible (see the last version of my code) but, for now, I'm only generating no name thanks the function `Bindlib.name_of`.

Reply:
Using Bindlib.name_of will create name clashes. So this is not sufficient. But this seems to go in the right direction. The problem is the generation of p_term's. We need to generate a unique fresh name for each tvar. Currently, what we were trying to do if generating the strings p0 .. pn from the tvar's x0 .. xn, where p is a pre-computed safe prefix. We can continue to do this. I found the following solution:
```
let fold_cons_typ (pos : popt)
      (inj_var : tvar list -> tvar -> 'term)
      (codom : term list -> 'term list -> 'term -> 'c)
      (build_rec_hyp : term list -> 'term -> 'term)
      (domrec : term -> tvar  -> 'term -> 'c -> 'c)
      (dom : term -> tvar -> 'c -> 'c)
      (ind_sym : sym) (cons_sym : sym)
      (f_rec : 'term -> 'term -> 'term -> 'term)
      (f_not_rec : 'term -> 'term -> 'term)
      (init : 'term) (s : string) : 'c =
  let rec aux : tvar list -> 'term list -> 'term -> term -> 'c =
    fun xs xs' acc a ->
    match Basics.get_args a with
    | (Symb(s), ts) ->
        if s == ind_sym then codom ts xs' acc
       else fatal pos "%a is not a constructor of %a"
              pp_symbol cons_sym pp_symbol ind_sym
    | (Prod(a,b), _) ->
        let (x,b) = Basics.unbind_name b s in
        let x' = inj_var xs x in
       begin
         match Basics.get_args a with
         | (Symb(s), ts) ->
             if s == ind_sym then
               let rec_hyp = build_rec_hyp ts x' in
               let next = aux (x::xs) (x'::xs') (f_rec acc x' rec_hyp) b in
               domrec a x rec_hyp next
             else
               let next = aux (x::xs) (x'::xs') (f_not_rec acc x') b in
               dom a x next
         | _ -> fatal pos "The type of %a is not supported" pp_symbol cons_sym
       end
    | _ -> fatal pos "The type of %a is not supported" pp_symbol cons_sym
  in aux [] [] init !(cons_sym.sym_type)


  let case_of : sym -> tbox = fun cons_sym ->
    (* 'term = tbox, 'c = tbox *)
    let inj_var _ x = _Vari x in
    let codom ts xs _ =
      prf_of_p (List.map lift ts) (fapp cons_sym (List.rev xs))
    in
    let build_rec_hyp ts x = prf_of_p (List.map lift ts) x in
    let domrec a x rec_hyp next =
      _Prod (lift a) (Bindlib.bind_var x (_Impl rec_hyp next))
    in
    let dom a x next = _Prod (lift a) (Bindlib.bind_var x next) in
    let f_rec a _ _ = a in
    let f_not_rec a _ = a in
    let acc = _Type in
    fold_cons_typ pos inj_var codom build_rec_hyp domrec dom ind_sym
      cons_sym f_rec f_not_rec acc x_str
  

     let gen_rule_cons : sym -> p_rule = fun cons_sym ->
    (* 'term = p_term, 'c = p_rule *)
    let pos : popt = None in
    let inj_var : tvar list -> tvar -> p_term = fun xs _ ->
      P.patt0 ("x" ^ string_of_int (List.length xs))
    in
    let codom ts xs p =
      let f (k, l) _ = (k+1, P.patt0 ("x" ^ string_of_int k) :: l) in
      let xs = snd (List.fold_left f (0, []) xs) in
      let rec_arg = P.fold_appl (P.iden cons_sym.sym_name) xs in
      let lhs = P.appl (P.appl_wild common_head (List.length ts)) rec_arg in
      if !log_enabled then
        log_ind "The rule [%a] --> %a"
          Pretty.pp_p_term lhs Pretty.pp_p_term p;
      P.rule lhs p
    in
    let build_rec_hyp ts x =
      let arg_type = P.appl_wild common_head (List.length ts) in
      P.appl arg_type x
    in
    let domrec _ _ _ next = next in
    let dom _ _ next = next in
    let f_rec p x rec_hyp = P.appl (P.appl p x) rec_hyp in
    let f_not_rec p x = P.appl p x in
    let acc = P.patt0 ("pi" ^ cons_sym.sym_name) in
    fold_cons_typ pos inj_var codom build_rec_hyp domrec dom ind_sym
      cons_sym f_rec f_not_rec acc ""
```

In the last part of the code, "x" should be replaced by a safe prefix.

Reply:
Your code doesn't compile.
And add an other argument for `aux` isn't readable.

Reply:
Well, it is compiling on my machine... Be careful that I permuted inj_var and codom, and added some arguments.

Reply:
Ok, I will try to modify this function myself one day later perhaps.

Reply:
There a solution of mine, but the names generated by Bindlib are empty and I didn't understand why.